### ISSUE TITLE
Bug: Annotated optional query parameter inconsistency

### ISSUE THREAD

**Issue by gsakkis** [litestar-org/litestar#2514]
### Description

Annotated optional query parameters are inconsistent compared to non annotated and required query parameters.

### MCVE

```python
from typing import Annotated

from litestar import Litestar, get
from litestar.params import Parameter


@get("/optional", sync_to_thread=False)
def optional(key: str | None = None) -> dict[str, str | None]:
    return {"key": key}


@get("/annotated_required", sync_to_thread=False)
def annotated_required(param: Annotated[str, Parameter(query="key")]) -> dict[str, str]:
    return {"key": param}


@get("/annotated_optional", sync_to_thread=False)
def annotated_optional(
    param: Annotated[str | None, Parameter(query="key")] = None
) -> dict[str, str | None]:
    return {"key": param}


app = Litestar([optional, annotated_required, annotated_optional])
```

### Steps to reproduce

```bash
# optional - non annotated
$ curl -w '\n' -s 'localhost:8000/optional?param=a'
{"key":null}
$ curl -w '\n' -s 'localhost:8000/optional?key=a'
{"key":"a"}

# annotated - non optional
$ curl -w '\n' -s 'localhost:8000/annotated_required?param=a'
{"status_code":400,"detail":"Missing required query parameter 'key' for url http://localhost:8000/annotated_required?param=a"}
$ curl -w '\n' -s 'localhost:8000/annotated_required?key=a'
{"key":"a"}

# annotated - optional
$ curl -w '\n' -s 'localhost:8000/annotated_optional?param=a'
{"key":"a"}
$ curl -w '\n' -s 'localhost:8000/annotated_optional?key=a'
{"key":null}
```
```


### Screenshots

```bash
"![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)"
```

### Litestar Version

2.2.1

### Platform

- [ ] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2514">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2514/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2514/pledge.svg">
</picture>
</a>

**Comment by gsakkis** [comment]
One more case that works consistently/correctly: Annotated optional without default:
```py
@get("/annotated_optional_no_default", sync_to_thread=False)
def annotated_optional_no_default(
    param: Annotated[str | None, Parameter(query="key")]
) -> dict[str, str | None]:
    return {"key": param}
```

So the (only?) odd one out is the annotated optional with default.

**Comment by peterschutt** [comment]
So to confirm, the inconsistency is that the request `/optional?param=a` returns `{"key": null}` while `/annotated_optional?param=a` returns `{"key": "a"}`. Have I got that right?

The inconsistency only seems to occur if the default value is `None`.

```py
    # same as yours except changed the default to a truthy value
    @get("/annotated_optional", sync_to_thread=False)
    def annotated_optional(param: Annotated[Optional[str], Parameter(query="key")] = "b") -> Dict[str, Optional[str]]:
        return {"key": param}

...
assert client.get("/annotated_optional?param=a").json() == {"key": "b"}  # True
```

**Comment by peterschutt** [comment]
Hmm

given:

```py
    @get("/optional-annotated-default")
    def handle_default_annotated(
        param: Annotated[Optional[str], Parameter(query="key")] = None
    ) -> Dict[str, Optional[str]]:
        return {"key": param}
```
then:

```py
>       assert optional_default_client.get("/optional-annotated-default", params={"key": "a"}).json() == {"key": "a"}
E       AssertionError: assert {'key': None} == {'key': 'a'}
E         Differing items:
E         {'key': None} != {'key': 'a'}
E         Use -v to get more diff
```
Seems like a value can't be provided at all for the optional/annotated/with default case..

**Comment by peterschutt** [comment]
What python version @gsakkis?

**Comment by gsakkis** [comment]
@peterschutt Python 3.10.12

**Comment by peterschutt** [comment]
Closed by #2516.

Added tests in #2515.
